### ISSUE TITLE
[Bug]: Model double sizes after training. Ho to make FP16 for prediction? 

### ISSUE THREAD

**Issue by iliaNecrov** [flairNLP/flair#3380]
### Describe the bug

I have flair in productuion but problem is that it is working quite slow. I was trying ONNX, but unfortuanally it doesn't work with deberta-v3. I am still investigating, maybe I do something wrong. 

But today's problem that I came across a thing that doubles my model size after training. It is common thing for deberta (https://discuss.huggingface.co/t/why-is-uploaded-model-twice-the-size-of-actual-model/18782/7) because it was trained with mixed precision and should be initialized as:

`model = AutoModelForSequenceClassification.from_pretrained(model_name, torch_dtype=torch.float16).to(device)`

with `torch_dtype=torch.float16` but flair doesn't do so. So I am facing issue that hugging face `microsoft/deberta-v3-small` weights **286 MB** and the same model but fine-tuned with flair on NER task weights **574 MB**. 

So my question is:
1) How do I get model in normal size? With fp16 training and prediction.
2) And maybe you have some advice about how can I get the maximum speed out of a model?

### To Reproduce

```python
import os

os.environ['CUDA_VISIBLE_DEVICES'] = "3" #,2,3" 
os.environ['TRANSFORMERS_CACHE'] = '/home/user/cache'
os.environ['TRANSFORMERS_OFFLINE'] = "1"

import flair 
flair.set_seed(2)

from flair.datasets import CONLL_03
from flair.embeddings import WordEmbeddings, FlairEmbeddings, StackedEmbeddings, CharacterEmbeddings
from flair.models import SequenceTagger
from flair.trainers import ModelTrainer
from flair.datasets import ColumnCorpus
from flair.trainers import ModelTrainer
from flair.data import MultiCorpus
from flair.data import Sentence
from flair.embeddings import TransformerWordEmbeddings
import datetime
import torch
import json

column_format={0: "text",3: "ner"} 
corpus_03 = CONLL_03(base_path='data/conll-2003', column_format = column_format, label_name_map={
    'PER': 'PER',
    'LOC': 'O',
    'ORG': 'ORG',
    'MISC': 'O'
    # by renaming to 'O' this tag gets ignored
})

corpus = MultiCorpus([corpus_03, ...])

label_type = 'ner'
label_dict = corpus.make_label_dictionary(label_type=label_type)

embeddings = TransformerWordEmbeddings(model='microsoft/deberta-v3-small',
                                       layers="-1",
                                       subtoken_pooling="first",
                                       fine_tune=True,
                                       use_context=False,
                                       )

tagger = SequenceTagger(hidden_size=256,
                        embeddings=embeddings,
                        tag_dictionary=label_dict,
                        tag_type='ner',
                        use_crf=False,
                        use_rnn=False,
                        reproject_embeddings=False,
                        )

trainer = ModelTrainer(tagger, corpus)
trainer.fine_tune('models/0_13_deberta/',
                  learning_rate=5.0e-5,
                  mini_batch_size=32,
                  embeddings_storage_mode='cpu',
                    monitor_test = True,
                 train_with_dev = True,
                  max_epochs=3)
```

### Expected behavior

Model with the same size as a hugging face one.

### Screenshots

![image](https://github.com/flairNLP/flair/assets/96539972/41770dc5-7df5-450a-b11c-4ac4027ea1d5)
![image](https://github.com/flairNLP/flair/assets/96539972/ff7644ec-1400-4792-8a66-7cbc95386767)

##### Flair
0.13.0
##### Pytorch
2.1.0.dev20230807+cu121
##### Transformers
4.21.0
#### GPU
True

**Comment by helpmefindaname** [comment]
I see, I only tested for creation & inference but for training it is a bit more complicated. 
The right thing would be to set the default float to 16bit, create the embeddings & model, set the default float back to 32 bit and then create the trainer. However that leads to a nan loss, as the 16bit weights is not stable enough for training.

I would suggest to instead create a 32bit model, but then use mixed precision training (setting `use_amp=True` on the trainer.fine_tune method) that way you will create a 32bit model but still train with 16bit computations. That makes the training stable again.
You can then still save the 16bit model, by running `tagger.half()` and then save it again via `tagger.save("model_path/model_name.pt")` 
Notice that converting the 32bit model to 16bit, might still reduce the model quality, so I recommend evaluating it again by running `print(tagger.evaluate(corpus.test, "ner").detailed_results)`

Interestingly, evaluating the 16bit model over the 32bit model did not only result into less memory requirements, but also increased the F1 score from 92.52% to 93.79%

So the full script for training is:
<details>

```python

import flair

flair.set_seed(2)

from flair.datasets import CONLL_03
from flair.models import SequenceTagger
from flair.trainers import ModelTrainer
from flair.embeddings import TransformerWordEmbeddings


def main():
    column_format = {0: "text", 3: "ner"}
    corpus_03 = CONLL_03(column_format=column_format, label_name_map={
        'PER': 'PER',
        'LOC': 'O',
        'ORG': 'ORG',
        'MISC': 'O'
        # by renaming to 'O' this tag gets ignored
    })

    corpus = corpus_03

    label_type = 'ner'
    label_dict = corpus.make_label_dictionary(label_type=label_type)

    embeddings = TransformerWordEmbeddings(model='microsoft/deberta-v3-small',
                                           layers="-1",
                                           subtoken_pooling="first",
                                           fine_tune=True,
                                           use_context=False,
                                           )

    tagger = SequenceTagger(hidden_size=256,
                            embeddings=embeddings,
                            tag_dictionary=label_dict,
                            tag_type='ner',
                            use_crf=False,
                            use_rnn=False,
                            reproject_embeddings=False,
                            )

    trainer = ModelTrainer(tagger, corpus)
    trainer.fine_tune('models/0_13_deberta/',
                      learning_rate=5.0e-5,
                      mini_batch_size=32,
                      embeddings_storage_mode='cpu',
                      monitor_test=True,
                      train_with_dev=True,
                      max_epochs=3,
                      use_amp=True,
                      )
    tagger.half()
    print(tagger.evaluate(corpus.test, "ner").detailed_results)
    tagger.save("models/0_13_deberta/model.pt")


if __name__ == "__main__":
    main()
```

</details>

for inference you then still need to do set the default 16 bit to load the model in an effective manner:

```python
import torch

from flair.datasets import CONLL_03
from flair.models import SequenceTagger


def main():
    column_format = {0: "text", 3: "ner"}
    corpus = CONLL_03(column_format=column_format, label_name_map={
        'PER': 'PER',
        'LOC': 'O',
        'ORG': 'ORG',
        'MISC': 'O'
        # by renaming to 'O' this tag gets ignored
    })
    torch.set_default_dtype(torch.float16)
    tagger = SequenceTagger.load("model.pt")
    torch.set_default_dtype(torch.float32)
    print(tagger.evaluate(corpus.test, "ner").detailed_results)


if __name__ == "__main__":
    main()

```

**Comment by iliaNecrov** [comment]
@helpmefindaname thanks a lot for your detailed explanation! 

One last thing I think that `torch.set_default_dtype(torch.float32)` line in your last code is not needed. We load model with fp16, but if setting default type to 32, we are back to 32. Am I wrong?

**Comment by helpmefindaname** [comment]
the setting the default back to 32bit is just there to not impact other models that might be corrupted by that.

But as you are saying it works, I will close this issue

**Comment by iliaNecrov** [comment]
@helpmefindaname yes, thanks again.
